### DISCUSSION TITLE
Artisan Schedule:run

### DISCUSSION THREAD

**Discussion by markkimsal** [tenancy/tenancy#303] [General]
I have 2 crons running `artisan schedule:run` and `artisan schedule:run --tenant=whitelabel1` .  I don't have dynamic tenants like one for each registered user or group.  There's basically just 2 tenants for the whole system and I keep track mostly by hostname.

When I run the schedule with the --tenant=whitelabel1, none of the commands get the --tenant= option passed along.  This caught me by surprise.

There's no `$this->options('tenant')` in the in the `Console/Kernel.php` file because that isn't the command that's run from `schedule:run`, it's the thing that defines the schedule for other artisan commands.

So, I had to hack in `$tenantOption = ' --tenant=whitelabel1'` for the run where my tenant has changed and append it to every command.  I feel like this wasn't really discussed anywhere in all the "affects X" portions of the documentation.  If it did, I missed it.

Is this a bad solution, how else would you go about it?  I'm not certain that I can get the `ResolvesTenants` in the Console/Kernel via DI to handle this dynamically and have only 1 cron command.  I tested it all with `schedule:list --tenant=xxx` and it looks right.

```php
    protected function schedule(Schedule $schedule): void
    {
        $tenantEnvironment = Tenancy::getFacadeRoot(Environment::class);
        $t = $tenantEnvironment->getTenant();
        $tenantOption = ' --tenant='. $t->getTenantKey();

        $schedule->command('app:create-sync-jobs' . $tenantOption)
            ->daily();
        /// repeat more times
 ```

**Comment by ArlonAntonius** [comment]
@markkimsal
Thank you for letting us know. I had a quick look at one of my own applications and using the schedule with the tenant should work flawlessly as long as you setup Console based identification :)

You can follow the documentation at: https://tenancy.dev/docs/tenancy/2.x/identification-console

And I have a small setup example here as well:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Symfony\Component\Console\Input\InputInterface;
use Tenancy\Identification\Concerns\AllowsTenantIdentification;
use Tenancy\Identification\Contracts\Tenant as ContractsTenant;
use Tenancy\Identification\Drivers\Console\Contracts\IdentifiesByConsole;

class Tenant extends Model implements ContractsTenant, IdentifiesByConsole
{
    use HasFactory, AllowsTenantIdentification;

    public function tenantIdentificationByConsole(InputInterface $input): ?ContractsTenant
    {
        if ($input->hasParameterOption('--tenant')) {
            return $this->query()
                ->find($input->getParameterOption('--tenant'));
        }

        return null;
    }
}
```

**Reply by markkimsal** [reply]
I have all that, but the tasks spawned from $scheduler->command() don't inherit that tenant param.  all console commands work fine, like migrate:status --tenant=xxx .  But I'm getting twice as many sync jobs on the main retail tenant and 0 sync jobs on the whitelabel tenant.  I wouldn't be able to pass --tenant=xxx down to the scheduled jobs if the command schedule:run wasn't hooked in correctly to IdentifiesByConsole.

I checked everything with `schedule:list` and `schedule:list --tenant=xxx` and the option is passed correctly when I append the string $tenantOption.

Is it something with spawning artisan commands, like internal laravel heartbeat vs external cron?

**Reply by ArlonAntonius** [reply]
@markkimsal I found exactly what you meant now.

It has to do with the fact that Laravel does not execute the command internally, it starts a completely new process to execute this and loses the parameters along the way.

I have an idea to temporarily fix this, but it would be highly opinionated and not in the light of how this ecosystem is supposed to work.

Within what timeframe would you need a solution for this?

**Reply by markkimsal** [reply]
I was just wondering if my solution of appending a string "--tenant=..." to the commands was normal/canonical.  It's working for me atm.

I had 2 thoughts:

1.   if you do `artisan schedule:work` does it still spawn external artisan commands?  does keeping artisan in the foreground somehow affect the command execution?
2.   Is there a way to cycle through every known tenant? I see `ResolvesTenants` in one example that seems like it might cycle through different tenant models until one is found.  (https://tenancy.dev/docs/tenancy/2.x/identification-console#custom-command)


For now, I have 2 or 3 cron entires in my /etc/cron.d/laravel-artisan-scheduler
```
* * * * * www-data cd $WORK_DIR && /usr/local/bin/php artisan schedule:run >> /dev/null 2>&1
* * * * * www-data cd $WORK_DIR && /usr/local/bin/php artisan schedule:run --tenant whitelabel1 >> /dev/null 2>&1
* * * * * www-data cd $WORK_DIR && /usr/local/bin/php artisan schedule:run --tenant whitelabel2 >> /dev/null 2>&1
```

**Comment by markkimsal** [comment]
There does not seem to be a nice way to influence the spawned command's environment or cli flags.

Inside Laravel scheduling event:
```php
    protected function execute($container)
    {
        return Process::fromShellCommandline(
            $this->buildCommand(), base_path(), null, null, null
        )->run();
    }
```
It looks to to leave any overriding env null when launching the Symfony Process.  So, there doesn't seem to be a nice way to hook into the `IdentifiesByEnv` path as a possible automatic solution.


Now I found `Illuminate\Console\Events\ScheduledTaskStarting` which could be listened to.  This event gets the $event (task) and still has time to modify the options.  But, it's not obvious that every scheduled artisan command would want that.

I guess I'm thinking about some sort of `AffectsScheduledCommands` and some kind of `ScheduledCommandsResolver` that can give the app owner discretion over which tasks need modifying.

But, in the end, I'm totally fine just querying the running process for resolved tenants and appending a string or $param [] with extra tenant information.
